bioptim/limits/constraints.py

```diff
@@ -299,6 +307,194 @@ def time_constraint(_: Constraint, all_pn: PenaltyNodeList, **unused_param):
+                    f" for such as dynamic: {dynamic_name}. "
+                    "It is only implemented for these dynamics:"
+                    " TORQUE_DRIVEN and TORQUE_DERIVATIVE_DRIVEN."
+                )
+
+            return BiorbdInterface.mx_to_cx("ForwardDynamics", qddot - qddot_fd, *var)
+
+        @staticmethod
+        def tau_equals_inverse_dynamics(_: Constraint, all_pn: PenaltyNodeList, **unused_param):
+            """
+            Compute the difference between symbolic joint torques and inverse dynamic results
+            It does not include any inversion of mass matrix
+
+            Parameters
+            ----------
+            _: Constraint
+                The actual constraint to declare
+            all_pn: PenaltyNodeList
+                The penalty node elements
+            **unused_param: dict
+                Since the function does nothing, we can safely ignore any argument
+            """
+
+            nlp = all_pn.nlp
+            q = nlp.states["q"].mx
+            qdot = nlp.states["qdot"].mx
+            tau = nlp.states["tau"].mx if "tau" in nlp.states.keys() else nlp.controls["tau"].mx
+            qddot = nlp.states["qddot"].mx if "qddot" in nlp.states.keys() else nlp.controls["qddot"].mx
+
+            if nlp.external_forces:
+                raise NotImplementedError(
+                    "This implicit constraint tau_equals_inverse_dynamics is not implemented yet with external forces"
+                )
+                # Todo: add fext tau_id = nlp.model.InverseDynamics(q, qdot, qddot, fext).to_mx()
+                # fext need to be a mx
+
+            tau_id = nlp.model.InverseDynamics(q, qdot, qddot).to_mx()
+            var = [nlp.states["q"], nlp.states["qdot"]]
+            dynamic_name = nlp.dynamics_type.type.name
+            if dynamic_name == "TORQUE_DERIVATIVE_DRIVEN":
+                var.extend(
+                    [
+                        nlp.states["tau"],
+                        nlp.controls["taudot"],
+                        nlp.states["qddot"],
+                        nlp.controls["qdddot"],
+                    ]
+                )
+
+            elif dynamic_name == "TORQUE_DRIVEN":
+                var.extend(
+                    [
+                        nlp.controls["tau"],
+                        nlp.controls["qddot"],
+                    ]
+                )
+            else:
+                raise NotImplementedError(
+                    "Implicit Dynamics has not been implemented yet"
+                    f" for such as dynamic: {dynamic_name}. "
+                    "It is only implemented for these dynamics:"
+                    " TORQUE_DRIVEN and TORQUE_DERIVATIVE_DRIVEN."
+                )
+
+            return BiorbdInterface.mx_to_cx("InverseDynamics", tau_id - tau, *var)
+
+        @staticmethod
+        def implicit_soft_contact_forces(_: Constraint, all_pn: PenaltyNodeList, **unused_param):
```

Comment:
Is this the same as `penalty.minimize_soft_contact_forces`? If so, there is no reason to double implement the same function, only implement it in penalty so it is usable by both objective and constraints

Reply:
This is not the same. here we enforce the contact force to be equals on contact force controls at each shooting nodes.
minimize_soft_contact_forces.

The part where i get contact forces is indeed the same where could i put this ?

Reply:
If  you were to put the contact force controls as the target, would they be the same?